Fix trunc instruction

kariya-mitsuru opened the pull request:
The instruction `inn.trunc_fmm_sx` should cause trap if source value
cannot be represented by target type.

Note: The old implementations of cast functions are renamed and
preserved for future use.  (Wasm 1.1 has the instruction
`inn.trunc_sat_fmm_sx` that should use these functions.)


Passed new 75 tests.

**before**
```
End ".../wain/spec-test/wasm-testsuite/conversions.wast":
  total: 436, passed: 369, failed: 67, skipped: 0

End ".../wain/spec-test/wasm-testsuite/traps.wast":
  total: 37, passed: 29, failed: 8, skipped: 0

Results of 76 files:
  total: 19740, passed: 19046, failed: 498, skipped: 196
```

**after**
```
End ".../wain/spec-test/wasm-testsuite/conversions.wast":
  total: 436, passed: 436, failed: 0, skipped: 0

End ".../wain/spec-test/wasm-testsuite/traps.wast":
  total: 37, passed: 37, failed: 0, skipped: 0

Results of 76 files:
  total: 19740, passed: 19121, failed: 423, skipped: 196
```

kariya-mitsuru committed:
Fix trunc instruction

The instruction inn.trunc_fmm_sx should cause trap if source value
cannot be represented by target type.

Note: The old implementations of cast functions are renamed and
preserved for future use.  (Wasm 1.1 has the instruction
inn.trunc_sat_fmm_sx that should use these functions.)

kariya-mitsuru commented:
すみません、一応この形でプルリクしたのですが、`float_to_int` マクロが条件を自動判別することに注力しすぎて、あまりにも技巧的過ぎる気もしています…
マクロの引数を 1 つ増やして手動で選択する方法もあるのですが、そちらの方がいいですかね…

rhysd commented:
ありがとうございます！やらないといけないと思いつつそのままになってた箇所なので助かります．明日コードを見させてください．

rhysd commented on wain-exec/src/cast.rs:
```diff
@@ -2,10 +2,61 @@
+                Some(f as $int)
+            } else {
+                None
+            }
```
マクロを1度しか使ってないので，マクロにしなくて良いかなと思います．

```rust
let gt_min = if MIN != MIN - 1.0 {
    MIN - 1.0 < f
} else {
    MIN <= f
};

if gt_min && f < MAX + 1.0 {
    Some(f as $int)
} else {
    None
}
```

rhysd commented on wain-exec/src/cast.rs:
```diff
@@ -22,103 +73,275 @@ macro_rules! float_to_int {
     use std::{f32, f64};
 
+    trait NextAfter {
+        fn nextafter(self, _: Self) -> Self;
```
単語の間は `_` で区切っておいていただけると．

```suggestion
        fn next_after(self, _: Self) -> Self;
```

rhysd reviewed (changes requested):
すみません，昨日プルリク全部見きれなかったので，今日になってしまいました．

kariya-mitsuru committed:
Follow Github comments

Remove unnecessary macro 'lt'.
Change function name 'nextafter' to 'next_after'

rhysd commented on wain-exec/src/cast.rs:
```diff
@@ -22,103 +73,275 @@ macro_rules! float_to_int {
+        fn nextafter(self, o: f32) -> f32 {
+            let b = self.to_bits();
+            f32::from_bits(if self.abs() < o.abs() { b + 1 } else { b - 1 })
+        }
```
これは `o` は origin で，origin 方向に 1bit 分近い数を返すメソッドで良いですよね？

kariya-mitsuru replied:
はい、そうです。
ちょっと（だいぶ）手抜きですが、テスト用ならこれぐらいでいいかな、と…

kariya-mitsuru replied:
ホントは Rust にホンモノの [`nextafter`](https://cpprefjp.github.io/reference/cmath/nextafter.html) があったら使いたかったんですが、どうもなさそうだったので…

rhysd replied:
あ，なるほど math.h 由来の名前でしたか．はい，挙動の確認をしたかっただけなので，実装的にはこれで問題ないと思います :+1:

rhysd commented on wain-exec/src/cast.rs:
```diff
@@ -2,10 +2,57 @@
+            //    b) If MIN - 1.0 can not be represented exactly,
+            //       then MIN - 1.0 = MIN.  (Because of "round to nearest even" behavior.)
+            //       So, we can check it with MIN <= f.
+            let gt_min = if MIN != MIN - 1.0 {
```
自分の理解を確認しておきたいのですが，`MIN != MIN - 1.0` が成り立つのって，ソース側 `fN` のビット幅 N がターゲット側 `iM` もしくは `uM` のビット幅 M より大きい（N > M）ときだけという理解で良いでしょうか？

kariya-mitsuru replied:
はい、だいたいそんな感じです。
より厳密に言うと、`iM` の場合は `fN::MANTISSA_DIGITS >= M` の場合で、`uM` の場合は常に成り立ちます。（`MIN` が `0` なので）
ただ、`M` と `N` はどうせ `32` と `64` しかないので `iM` の場合は実質同じです。

rhysd replied:
そうか，fraction の幅が対象より広ければ常に正確に表せるので成り立ちますね．unsigned の場合もおっしゃる通りでした．ありがとうございます．実質 `f64 → i32` の場合限定ですね

rhysd commented:
ありがとうございます！
